### PR TITLE
Fix ONNX exports for Optimum compatible models

### PR DESCRIPTION
**Pull request by merveenoyan**
@amyeroberts as discussed and also pinging @xenova for review :') (who also fixed DPT) 

I prioritized Optimum compatible ones because I'm launching a project where there's Optimum examples for vision models. I will have a separate PR for the models that aren't compatible with Optimum. Rest of the Optimum compatible models export well without a problem.

### CODE REVIEW

File: src/transformers/models/dpt/modeling_dpt.py [hunk 1]
```diff
@@ -226,7 +226,9 @@ def _resize_pos_embed(self, posemb, grid_size_height, grid_size_width, start_ind
         posemb_tok = posemb[:, :start_index]
         posemb_grid = posemb[0, start_index:]
 
-        old_grid_size = int(math.sqrt(len(posemb_grid)))
+        def safe_int(x):
+          return x.to(torch.int64) if torch.jit.is_tracing() else int(x)
+        old_grid_size = safe_int(posemb_grid.size(0) ** 0.5)
```
Comment (xenova) [inline src/transformers/models/dpt/modeling_dpt.py]: It might be necessary to extract this cast function to a utility file, and then re-use across different models. I have a feeling that other instances, like:
image

will break execution, since during inference, it may require normal python types

File: src/transformers/utils/generic.py [hunk 2]
```diff
@@ -750,3 +750,15 @@ def infer_framework(model_class):
             return "flax"
     else:
         raise TypeError(f"Could not infer framework from class {model_class}.")
+
+
+def safe_int(x):
+    import torch
```
Comment (amyeroberts) [inline src/transformers/utils/generic.py:756]: We need pro

File: src/transformers/utils/generic.py [hunk 3]
```diff
@@ -750,3 +750,15 @@ def infer_framework(model_class):
...
+
+    return x.to(torch.int64) if torch.jit.is_tracing() else int(x)
+
+
+def safe_float(x):
+    import torch
+
+    return x.to(torch.float32) if torch.jit.is_tracing() else int(x)
```
Comment (amyeroberts) [inline src/transformers/utils/generic.py:764]: We need to protect the torch import here 

```suggestion
def safe_int(x):
    if not is_torch_available():
        return int(x)
        
    import torch

    return x.to(torch.int64) if torch.jit.is_tracing() else int(x)


def safe_float(x):
    if not is_torch_available():
        return int(x)

    import torch

    return x.to(torch.float32) if torch.jit.is_tracing() else int(x)
```

File: src/transformers/utils/generic.py [hunk 4]
```diff
@@ -750,3 +750,15 @@ def infer_framework(model_class):
             return "flax"
     else:
         raise TypeError(f"Could not infer framework from class {model_class}.")
+
+
+def safe_int(x):
```
Comment (amyeroberts) [inline src/transformers/utils/generic.py:755]: Docstrings would be helpful here e.g. for inspecting in IDEs: what does it mean for an int to be safe?

File: src/transformers/models/mobilevit/modeling_mobilevit.py [hunk 5]
```diff
@@ -437,8 +438,16 @@ def unfolding(self, features: torch.Tensor) -> Tuple[torch.Tensor, Dict]:
 
         batch_size, channels, orig_height, orig_width = features.shape
 
-        new_height = int(math.ceil(orig_height / patch_height) * patch_height)
-        new_width = int(math.ceil(orig_width / patch_width) * patch_width)
+        new_height = (
+            safe_float(torch.ceil(orig_height / patch_height) * patch_height)
+            if torch.jit.is_tracing()
```
Comment (amyeroberts) [inline src/transformers/models/mobilevit/modeling_mobilevit.py:443]: Do we need the conditional here? This is already handled in the `safe_float` and `safe_int` functions

**Review by amyeroberts — COMMENTED** [review]
Very nice! Thanks for fixing this for all these models ❤️ 

Just a few small comments

File: src/transformers/models/mobilevit/modeling_mobilevit.py (hunk 5 above)
Comment (merveenoyan) [reply src/transformers/models/mobilevit/modeling_mobilevit.py:443]: I think it's required for `torch.ceil` no?

Comment (amyeroberts) [reply src/transformers/models/mobilevit/modeling_mobilevit.py:443]: tbh, I don't know, is `torch.ceil` not tracing compatible?

Comment (merveenoyan) [reply src/transformers/models/mobilevit/modeling_mobilevit.py:443]: one needs to convert a data type to a tensor before using `torch.ceil` AFAIK

File: src/transformers/utils/generic.py (hunk 4 above)
Comment (xenova) [reply src/transformers/utils/generic.py:755]: Indeed a better name is probably a good idea 😅 I called it `safe_int` in a way to "safely cast some value (which could be a python number or tensor) to an integer in a way that respects tracing"

File: src/transformers/models/mobilevit/modeling_mobilevit.py (hunk 5 above)
Comment (merveenoyan) [reply src/transformers/models/mobilevit/modeling_mobilevit.py:443]: if I'm passing an int or float, torch.ceil will be called first and it will fail because torch.ceil can only be called with tensors AFAIK

Comment (amyeroberts) [reply src/transformers/models/mobilevit/modeling_mobilevit.py:443]: Makes sense!

Comment (amyeroberts) [reply src/transformers/models/mobilevit/modeling_mobilevit.py:443]: Only other Q here then is why do we use a `float` when tracing and int otherwise?

Comment (merveenoyan) [reply src/transformers/models/mobilevit/modeling_mobilevit.py:443]: sorry I think I was mistaken with that one, you're right, I fixed it :)

File: src/transformers/utils/generic.py (hunk 4 above)
Comment (merveenoyan) [reply src/transformers/utils/generic.py:756]: I'll swap with torch.int

**Comment by merveenoyan** [comment]
@amyeroberts the failing tests seem irrelevant to this PR, I can't re-run them, can you re-run?

**Comment by amyeroberts** [comment]
@merveenoyan si si - done!

**Review by amyeroberts — APPROVED** [review]
Thanks for fixing all of these! 

Just for my own understanding - it there any reason to not use the torch compatible float/int when were not tracing?

**Comment by merveenoyan** [comment]
@amyeroberts to my understanding, torch ONNX export internally calls `trace` thus to capture the graph properly all needs to be torch tensor, one needs to check tracing, and convert to torch tensor/convert ops internally if it's not torch

**Comment by merveenoyan** [comment]
@amyeroberts can you merge if you think it's ok?

**Comment by amyeroberts** [comment]
> @amyeroberts to my understanding, torch ONNX export internally calls trace thus to capture the graph properly all needs to be torch tensor, one needs to check tracing, and convert to torch tensor/convert ops internally if it's not torch

Right, I see why we need to do it for the onnx export, but for day-to-day use could we just use torch primitives instead of a python `int` or `float` i.e. do we need to maintain this if/else structure or can we switch everything to torch land regardless of whether we're tracing or not?

**Comment by merveenoyan** [comment]
@amyeroberts I guess if it's just torch modelling code then yes. Would you like me to swap everything?

**Comment by merveenoyan** [comment]
also asking the same question to @xenova

**Comment by amyeroberts** [comment]
> Would you like me to swap everything?

@merveenoyan Yes please! This will be cleaner and easier to follow in the code :)

**Comment by xenova** [comment]
> also asking the same question to @xenova

I agree with @amyeroberts - if there is a way to "do everything in torch land", that's the best solution! However, there are cases where I'm not entirely sure how to do this. For example, with `torch.nn.interpolate`:
- during inference, the sizes/scales CANNOT be tensors. They must be python ints/floats.
- during tracing, the sizes/scales SHOULD be tensors. Dynamic shapes usually break if they are python ints/floats due to the loss of tracing.

See [here](https://github.com/YvanYin/Metric3D/issues/117#issue-2364404569) for example code.

**Comment by merveenoyan** [comment]
@xenova sounds good, very glad to work with you tbh I didn't know that it would be required in inference.
@amyeroberts should we merge?

**Comment by amyeroberts** [comment]
@merveenoyan My understanding from above was that the PR would be updated to remove all the if/else structures wherever possible (but as @xenova points out isn't everywhere unfortunately)

**Comment by merveenoyan** [comment]
@amyeroberts from what I understood we should still keep them in if/else not to break the inference (I'm also scared of edge cases if there is etc) so I'd rather keep them. what I can do is to test all of them to see if they break or not when all are tensors and remove where it doesn't have to be a python type

**Comment by amyeroberts** [comment]
@merveenoyan OK. Let's just merge then and we can follow up in future PRs 👍
